pip/pip-269.md

```diff
@@ -0,0 +1,242 @@
+
+# Goals
+**Sub-Goal-1**: when a **Cursor Position Reset*** occurs, reject new reading, discard the in-flight reading and half in-flight reading<subp>[6]</sup>. And take care of these two things:
+- Ensure that changes to `read position` and `mark deleted position` are thread-safe.
+- Since more than one component uses Cursor to do **Sequential Read** or **Message Replay Read**, these components may also have their own attributes that need to be modified atomicity, we should also provide a mechanism to guarantee the atomicity modification of both attributes of Cursor and attributes of other components. 
+**Sub-Goal-2**: Remove some patchs(referred to as **Goal-3**):
+  - Revert https://github.com/apache/pulsar/issues/6403 (it named "A compensation mechanism of **Message Replay Read**" above)
+  - Remove the mechanism Rewind After In-flight Reading<sup>[5]</sup>
+**Out of scope**: Prevent more than one **Cursor Position Reset** executed at the same time(Current Proposal does not consider this optimization, there will be some bug fixes after this PIP)
+
+# High-Level Design
+
+### New concepts
+- **Cursor position is resetting**: indicate if there is a **Cursor Position Reset** running. Inject new readingg and discard in-flight reading if it is `true`.
+- **Cursor Epoch**: a counter to record how many times the the **Cursor Position Reset** executed. If it is increased after a in-flight reading, discard this reading.
+  - This attribute can help to prevent such error: a reading is started(`${Cursor position is resetting}` is `false`); a **Cursor position is resetting** is executed; the reading is completed(`${Cursor position is resetting}` is `false`).
+
+### Guarantee safe changes
+- Inside Cursor: the changes of attributes `read position`, `mark deleted position`, `cursor epoch`, and `cursor position is resetting` use the same lock.
+- Outside Cursor
+  - The action **Cursor Position Reset*** update `${Cursor position is resetting}` first, which guarantees that no other **Cursor Position Reset*** events will reenter(fast fail), and new reading can be rejected (In other words, it is the same as a non-reentrant lock).
+  - Each component (such as Dispatcher, Replicator...) modifies its attributes(each component to decide whether it needs to a locked).
+  - End the change: revert `${Cursor position is resetting}`; update `read position` and `mark deleted position`; increase **Cursor Epoch**(help to prevent such error: a reading is started(`${Cursor Epoch}` is increased); a **Cursor position is resetting** is executed; the reading is completed).
+  
+# Detailed Design
+
+**ManagedLedgerException.java**
+```java
+/** If a reading was injected by the event Cursor Position Reset, it will get a CursorModifyingException. **/
+public static class CursorModifyingException extends ManagedLedgerException {};
+```
+
+**CursorEpoch.java**
```

Comment:
Since it's all about cursor position reset, I think the name should signify that. not sure yet what it is.

Reply:
renamed to `CursorStateRevisionController `